onnxruntime/core/optimizer/transformer_memcpy.cc

```diff
@@ -274,22 +290,37 @@ static const onnxruntime::NodeArg* FindNodeArg(const NodeArgSetType& def_set, co
 // We duplicate any initializer that is used by both provider nodes and non-provider nodes
 // to ensure that provider nodes and non-provider nodes don't share initializers, as they
 // need to stay in different memory locations.
-void TransformerMemcpyImpl::ProcessInitializers(const KernelRegistryManager& kernel_registries) {
+bool TransformerMemcpyImpl::ProcessInitializers(const KernelRegistryManager& kernel_registries, const InitializedTensorSet& initializers_consumed) {
   std::map<const onnxruntime::NodeArg*, onnxruntime::NodeArg*> replacements;
-  for (const auto& pair : graph_.GetAllInitializedTensors()) {
+  for (const auto& pair : initializers_consumed) {
     const auto& name = pair.first;
     const onnxruntime::NodeArg* provider_def = FindNodeArg(provider_input_defs_, name);
     const onnxruntime::NodeArg* non_provider_def = FindNodeArg(non_provider_input_defs_, name);
     if (provider_def != nullptr && non_provider_def != nullptr) {
       std::string new_def_name = graph_.GenerateNodeArgName(name);
       auto& new_def = graph_.GetOrCreateNodeArg(new_def_name, provider_def->TypeAsProto());
 
-      const TensorProto* tensor_proto = nullptr;
-      bool found = graph_.GetInitializedTensor(name, tensor_proto);
-      ORT_ENFORCE(found, "Failed to get initialized tensor ", name);
+      const TensorProto* tensor_proto = pair.second;
+      ORT_ENFORCE(tensor_proto != nullptr, "Failed to get initialized tensor ", pair.first);
 
       TensorProto new_tensor_proto = *tensor_proto;
       *(new_tensor_proto.mutable_name()) = new_def_name;
+
+      // We make a copy of the initializer that is to be consumed by the provider Node so that
+      // session state initializer can copy it over to the provider device during its operation
+      LOGS(logger_, INFO) << "Making a copy of the initializer: " << name << " and the copy's name is: " << new_def_name;
+
+      if (graph_.IsSubgraph()) {
```

Comment:
this is not needed. Adding a TODO or NOTE comment here is good enough.

Reply:
If you have multiple subgraphs using the same initializer from the main graph on a non-cpu device, each will copy the initializer. If that's a large weight there could be significant impact to the memory usage. It's difficult to handle better and may not come up enough to warrant that effort. Logging it provides an easy way to identify that this could be the source of unexpected memory usage. The log happens during loading so doesn't impact execution time. 

What's the reason for not wanting to log such a thing?

Reply:
I got the point of trying to get attention that some copy may not be needed. Fair. then adding a TODO or even create a task tracking it. 

I would like to not have this logging, which does not add much value but overhead. The overhead may not be much, but, i'm being pushed more sensitive for some real cheap models from our customers. Let's not add more. Thank you!

Reply:
Sorry one more question. Given the log is at INFO level and our default is WARNING, it won't add cost unless someone changes the log level. Given that, is the guideline to still not have it? 

And is there a different guideline re: logging for code that happens during model load/initialization (like this) vs what happens during model execution? 

Maybe this info should be documented in the coding guidelines so it's more widely known.

Reply:
Removed all logging (kept comments) per offline discussion